File: internal/sstable/reader.go
package sstable

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"hash/crc32"
	"io"
	"os"
	"sort"

	"bigtable/internal/bloom"
	"bigtable/internal/record"
)

func Open(path string) (*File, error) {
	fh, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("open sstable: %w", err)
	}

	level, seq, ok := parseLevelAndSeq(path)
	if !ok {
		_ = fh.Close()
		return nil, fmt.Errorf("parse sstable name: %w", ErrCorruptFooter)
	}

	info, err := fh.Stat()
	if err != nil {
		_ = fh.Close()
		return nil, fmt.Errorf("stat sstable: %w", err)
	}

	var footer Footer
	switch {
	case info.Size() >= newFooterSize:
		footerBytes := make([]byte, newFooterSize)
		_, err = fh.ReadAt(footerBytes, info.Size()-newFooterSize)
		if err != nil {
			_ = fh.Close()
			return nil, fmt.Errorf("read footer: %w", err)
		}

		magic := string(bytes.TrimRight(footerBytes[32:40], "\x00"))
		if magic == newMagicValue {
			footer = Footer{
				IndexOffset: binary.BigEndian.Uint64(footerBytes[0:8]),
				IndexCount:  binary.BigEndian.Uint32(footerBytes[8:12]),
				RecordCount: binary.BigEndian.Uint32(footerBytes[12:16]),
				SparseGap:   binary.BigEndian.Uint32(footerBytes[16:20]),
				BloomOffset: binary.BigEndian.Uint64(footerBytes[20:28]),
				BloomSize:   binary.BigEndian.Uint32(footerBytes[28:32]),
				Magic:       newMagicValue,
			}
		} else {
			legacyFooterBytes := make([]byte, legacyFooterSize)
			_, err = fh.ReadAt(legacyFooterBytes, info.Size()-legacyFooterSize)
			if err != nil {
				_ = fh.Close()
				return nil, fmt.Errorf("read legacy footer: %w", err)
			}
			if string(legacyFooterBytes[20:28]) != legacyMagicValue {
				_ = fh.Close()
				return nil, ErrCorruptFooter
			}
			footer = Footer{
				IndexOffset: binary.BigEndian.Uint64(legacyFooterBytes[0:8]),
				IndexCount:  binary.BigEndian.Uint32(legacyFooterBytes[8:12]),
				RecordCount: binary.BigEndian.Uint32(legacyFooterBytes[12:16]),
				SparseGap:   binary.BigEndian.Uint32(legacyFooterBytes[16:20]),
				Magic:       legacyMagicValue,
			}
		}
	default:
		_ = fh.Close()
		return nil, ErrCorruptFooter
	}

	index, err := readIndex(fh, footer)
	if err != nil {
		_ = fh.Close()
		return nil, err
	}

	var bf *bloom.Filter
	if footer.Magic == newMagicValue && footer.BloomOffset > 0 && footer.BloomSize > 0 {
		bf, err = readBloom(fh, footer)
		if err != nil {
			_ = fh.Close()
			return nil, err
		}
	}

	return &File{
		path:   path,
		fh:     fh,
		index:  index,
		footer: footer,
		bloom:  bf,
		level:  level,
		seq:    seq,
	}, nil
}

func (f *File) Get(key []byte) (record.Record, bool, error) {
	if f == nil || f.fh == nil {
		return record.Record{}, false, fmt.Errorf("sstable closed")
	}

	if f.bloom != nil && !f.bloom.MightContain(key) {
		return record.Record{}, false, nil
	}

	startOffset := uint64(0)
	endOffset := f.footer.IndexOffset
	if f.footer.BloomOffset > 0 {
		endOffset = f.footer.BloomOffset
	}

	if len(f.index) > 0 {
		pos := sort.Search(len(f.index), func(i int) bool {
			return bytes.Compare(f.index[i].Key, key) > 0
		})

		candidate := pos - 1
		if candidate >= 0 {
			startOffset = f.index[candidate].Offset
			if candidate+1 < len(f.index) {
				endOffset = f.index[candidate+1].Offset
			}
		} else {
			startOffset = 0
			endOffset = f.index[0].Offset
		}
	}

	offset := startOffset
	for offset < endOffset {
		rec, nextOffset, err := readRecordAt(f.fh, offset, endOffset)
		if err != nil {
			return record.Record{}, false, err
		}

		cmp := bytes.Compare(rec.Key, key)
		if cmp == 0 {
			if rec.IsDeleted {
				return rec, true, nil
			}
			return rec, true, nil
		}
		if cmp > 0 {
			return record.Record{}, false, nil
		}

		offset = nextOffset
	}

	return record.Record{}, false, nil
}

func (f *File) All() ([]record.Record, error) {
	if f == nil || f.fh == nil {
		return nil, fmt.Errorf("sstable closed")
	}

	records := make([]record.Record, 0)
	offset := uint64(0)
	limit := f.footer.IndexOffset
	if f.footer.BloomOffset > 0 {
		limit = f.footer.BloomOffset
	}

	for offset < limit {
		rec, nextOffset, err := readRecordAt(f.fh, offset, limit)
		if err != nil {
			if err == io.EOF {
				break
			}
			return nil, err
		}

		records = append(records, rec)
		offset = nextOffset
	}

	return records, nil
}

func readBloom(fh *os.File, footer Footer) (*bloom.Filter, error) {
	data := make([]byte, footer.BloomSize)

	_, err := fh.ReadAt(data, int64(footer.BloomOffset))
	if err != nil {
		return nil, fmt.Errorf("read bloom filter: %w", err)
	}

	filter, ok := bloom.UnmarshalBinary(data)
	if !ok {
		return nil, ErrCorruptFooter
	}

	return filter, nil
}

func readIndex(fh *os.File, footer Footer) ([]SparseIndexEntry, error) {
	if footer.IndexCount == 0 {
		return nil, nil
	}

	index := make([]SparseIndexEntry, 0, footer.IndexCount)
	offset := int64(footer.IndexOffset)

	for i := uint32(0); i < footer.IndexCount; i++ {
		var keyLenBuf [4]byte
		_, err := fh.ReadAt(keyLenBuf[:], offset)
		if err != nil {
			return nil, fmt.Errorf("read index key len: %w", err)
		}
		offset += 4

		keyLen := binary.BigEndian.Uint32(keyLenBuf[:])
		key := make([]byte, keyLen)
		_, err = fh.ReadAt(key, offset)
		if err != nil {
			return nil, fmt.Errorf("read index key: %w", err)
		}
		offset += int64(keyLen)

		var offBuf [8]byte
		_, err = fh.ReadAt(offBuf[:], offset)
		if err != nil {
			return nil, fmt.Errorf("read index offset: %w", err)
		}
		offset += 8

		index = append(index, SparseIndexEntry{
			Key:    key,
			Offset: binary.BigEndian.Uint64(offBuf[:]),
		})
	}

	return index, nil
}

func readRecordAt(fh *os.File, offset uint64, limit uint64) (record.Record, uint64, error) {
	if offset+4 > limit {
		return record.Record{}, 0, io.EOF
	}

	var lenBuf [4]byte
	_, err := fh.ReadAt(lenBuf[:], int64(offset))
	if err != nil {
		return record.Record{}, 0, fmt.Errorf("read record length: %w", err)
	}

	rawLen := binary.BigEndian.Uint32(lenBuf[:])
	hasCRC := rawLen&checksumFlag != 0
	payloadLen := rawLen &^ checksumFlag

	total := uint64(4) + uint64(payloadLen)
	if hasCRC {
		total += 4
	}
	if offset+total > limit {
		return record.Record{}, 0, ErrCorruptRecord
	}

	payload := make([]byte, payloadLen)
	_, err = fh.ReadAt(payload, int64(offset+4))
	if err != nil {
		return record.Record{}, 0, fmt.Errorf("read record payload: %w", err)
	}

	if hasCRC {
		var crcBuf [4]byte
		_, err = fh.ReadAt(crcBuf[:], int64(offset+4+uint64(payloadLen)))
		if err != nil {
			return record.Record{}, 0, fmt.Errorf("read record crc: %w", err)
		}

		want := binary.BigEndian.Uint32(crcBuf[:])
		got := crc32.ChecksumIEEE(payload)
		if want != got {
			return record.Record{}, 0, ErrCorruptRecord
		}
	}

	rec, err := record.Decode(payload)
	if err != nil {
		return record.Record{}, 0, fmt.Errorf("decode record: %w", err)
	}

	return rec, offset + total, nil
}

File: internal/sstable/reader.cpp
#include "reader.h"
#include <filesystem>
#include <algorithm>
#include <iostream>
#include <cstring>

namespace fs = std::filesystem;

namespace bigdb::sstable {

namespace {

uint32_t crc32_ieee(const std::vector<uint8_t>& data) {
    uint32_t crc = 0xFFFFFFFF;
    for (uint8_t byte : data) {
        crc ^= byte;
        for (int i = 0; i < 8; i++) {
            uint32_t mask = -(crc & 1);
            crc = (crc >> 1) ^ (0xEDB88320 & mask);
        }
    }
    return ~crc;
}

std::string filepathBase(const std::string& path) {
    fs::path p(path);
    return p.filename().string();
}

bool parseLevelAndSeq(const std::string& path, uint32_t& level, uint64_t& seq) {
    std::string name = filepathBase(path);

    if (name.rfind("l", 0) == 0 && name.find("-sst-") != std::string::npos && name.length() >= 4 && name.compare(name.length() - 4, 4, ".sst") == 0) {
        size_t pos = name.find("-sst-");
        std::string levelRaw = name.substr(1, pos - 1);
        std::string seqRaw = name.substr(pos + 5, name.length() - pos - 5 - 4);

        try {
            level = std::stoul(levelRaw);
            seq = std::stoull(seqRaw);
            return true;
        } catch (...) {
            return false;
        }
    }

    if (name.rfind("sst-", 0) == 0 && name.length() >= 4 && name.compare(name.length() - 4, 4, ".sst") == 0) {
        std::string seqRaw = name.substr(4, name.length() - 4 - 4);
        try {
            level = 0;
            seq = std::stoull(seqRaw);
            return true;
        } catch (...) {
            return false;
        }
    }

    return false;
}

std::pair<record::Record, uint64_t> readRecordAt(std::ifstream& fh, uint64_t offset, uint64_t limit) {
    if (offset + 4 > limit) {
        throw std::runtime_error("EOF");
    }

    fh.seekg(offset, std::ios::beg);
    uint8_t lenBuf[4];
    if (!fh.read(reinterpret_cast<char*>(lenBuf), 4)) {
        throw std::runtime_error("read record length failed");
    }

    uint32_t rawLen = (static_cast<uint32_t>(lenBuf[0]) << 24) |
                      (static_cast<uint32_t>(lenBuf[1]) << 16) |
                      (static_cast<uint32_t>(lenBuf[2]) << 8) |
                      static_cast<uint32_t>(lenBuf[3]);

    bool hasCRC = (rawLen & ChecksumFlag) != 0;
    uint32_t payloadLen = rawLen & ~ChecksumFlag;

    uint64_t total = 4 + payloadLen;
    if (hasCRC) total += 4;

    if (offset + total > limit) {
        throw CorruptRecordError();
    }

    std::vector<uint8_t> payload(payloadLen);
    if (!fh.read(reinterpret_cast<char*>(payload.data()), payloadLen)) {
        throw std::runtime_error("read record payload failed");
    }

    if (hasCRC) {
        uint8_t crcBuf[4];
        if (!fh.read(reinterpret_cast<char*>(crcBuf), 4)) {
            throw std::runtime_error("read record crc failed");
        }
        uint32_t want = (static_cast<uint32_t>(crcBuf[0]) << 24) |
                        (static_cast<uint32_t>(crcBuf[1]) << 16) |
                        (static_cast<uint32_t>(crcBuf[2]) << 8) |
                        static_cast<uint32_t>(crcBuf[3]);
        uint32_t got = crc32_ieee(payload);
        if (want != got) {
            throw CorruptRecordError();
        }
    }

    record::Record rec = record::Record::decode(payload);
    return {rec, offset + total};
}

std::unique_ptr<bloom::Filter> readBloom(std::ifstream& fh, const Footer& footer) {
    std::vector<uint8_t> data(footer.bloom_size);
    fh.seekg(footer.bloom_offset, std::ios::beg);
    if (!fh.read(reinterpret_cast<char*>(data.data()), footer.bloom_size)) {
        throw std::runtime_error("read bloom filter failed");
    }

    auto filter = bloom::Filter::UnmarshalBinary(data);
    if (!filter) {
        throw CorruptFooterError();
    }
    return filter;
}

std::vector<SparseIndexEntry> readIndex(std::ifstream& fh, const Footer& footer) {
    if (footer.index_count == 0) {
        return {};
    }

    std::vector<SparseIndexEntry> index;
    index.reserve(footer.index_count);
    uint64_t offset = footer.index_offset;

    for (uint32_t i = 0; i < footer.index_count; i++) {
        fh.seekg(offset, std::ios::beg);
        uint8_t keyLenBuf[4];
        if (!fh.read(reinterpret_cast<char*>(keyLenBuf), 4)) {
            throw std::runtime_error("read index key len failed");
        }
        offset += 4;

        uint32_t keyLen = (static_cast<uint32_t>(keyLenBuf[0]) << 24) |
                          (static_cast<uint32_t>(keyLenBuf[1]) << 16) |
                          (static_cast<uint32_t>(keyLenBuf[2]) << 8) |
                          static_cast<uint32_t>(keyLenBuf[3]);

        std::vector<uint8_t> key(keyLen);
        if (!fh.read(reinterpret_cast<char*>(key.data()), keyLen)) {
            throw std::runtime_error("read index key failed");
        }
        offset += keyLen;

        uint8_t offBuf[8];
        if (!fh.read(reinterpret_cast<char*>(offBuf), 8)) {
            throw std::runtime_error("read index offset failed");
        }
        offset += 8;

        uint64_t off = (static_cast<uint64_t>(offBuf[0]) << 56) |
                       (static_cast<uint64_t>(offBuf[1]) << 48) |
                       (static_cast<uint64_t>(offBuf[2]) << 40) |
                       (static_cast<uint64_t>(offBuf[3]) << 32) |
                       (static_cast<uint64_t>(offBuf[4]) << 24) |
                       (static_cast<uint64_t>(offBuf[5]) << 16) |
                       (static_cast<uint64_t>(offBuf[6]) << 8) |
                       static_cast<uint64_t>(offBuf[7]);

        index.push_back({key, off});
    }

    return index;
}

} // namespace

File::File(const std::string& path, std::unique_ptr<std::ifstream> fh, std::vector<SparseIndexEntry> index, Footer footer, std::unique_ptr<bloom::Filter> bloom, uint32_t level, uint64_t seq)
    : path_(path), fh_(std::move(fh)), index_(std::move(index)), footer_(footer), bloom_(std::move(bloom)), level_(level), seq_(seq) {}

File::~File() {
    Close();
}

void File::Close() {
    if (fh_) {
        fh_->close();
        fh_.reset();
    }
}

std::unique_ptr<File> File::Open(const std::string& path) {
    auto fh = std::make_unique<std::ifstream>(path, std::ios::binary);
    if (!*fh) {
        throw std::runtime_error("open sstable failed");
    }

    uint32_t level = 0;
    uint64_t seq = 0;
    if (!parseLevelAndSeq(path, level, seq)) {
        throw CorruptFooterError();
    }

    fh->seekg(0, std::ios::end);
    size_t size = fh->tellg();

    Footer footer;
    if (size >= NewFooterSize) {
        fh->seekg(size - NewFooterSize, std::ios::beg);
        std::vector<uint8_t> footerBytes(NewFooterSize);
        fh->read(reinterpret_cast<char*>(footerBytes.data()), NewFooterSize);

        std::string magic(reinterpret_cast<char*>(footerBytes.data() + 32), 8);
        magic.erase(std::find(magic.begin(), magic.end(), '\0'), magic.end());
        
        if (magic == NewMagicValue) {
            footer.index_offset = (static_cast<uint64_t>(footerBytes[0]) << 56) | (static_cast<uint64_t>(footerBytes[1]) << 48) | (static_cast<uint64_t>(footerBytes[2]) << 40) | (static_cast<uint64_t>(footerBytes[3]) << 32) | (static_cast<uint64_t>(footerBytes[4]) << 24) | (static_cast<uint64_t>(footerBytes[5]) << 16) | (static_cast<uint64_t>(footerBytes[6]) << 8) | static_cast<uint64_t>(footerBytes[7]);
            footer.index_count = (static_cast<uint32_t>(footerBytes[8]) << 24) | (static_cast<uint32_t>(footerBytes[9]) << 16) | (static_cast<uint32_t>(footerBytes[10]) << 8) | static_cast<uint32_t>(footerBytes[11]);
            footer.record_count = (static_cast<uint32_t>(footerBytes[12]) << 24) | (static_cast<uint32_t>(footerBytes[13]) << 16) | (static_cast<uint32_t>(footerBytes[14]) << 8) | static_cast<uint32_t>(footerBytes[15]);
            footer.sparse_gap = (static_cast<uint32_t>(footerBytes[16]) << 24) | (static_cast<uint32_t>(footerBytes[17]) << 16) | (static_cast<uint32_t>(footerBytes[18]) << 8) | static_cast<uint32_t>(footerBytes[19]);
            footer.bloom_offset = (static_cast<uint64_t>(footerBytes[20]) << 56) | (static_cast<uint64_t>(footerBytes[21]) << 48) | (static_cast<uint64_t>(footerBytes[22]) << 40) | (static_cast<uint64_t>(footerBytes[23]) << 32) | (static_cast<uint64_t>(footerBytes[24]) << 24) | (static_cast<uint64_t>(footerBytes[25]) << 16) | (static_cast<uint64_t>(footerBytes[26]) << 8) | static_cast<uint64_t>(footerBytes[27]);
            footer.bloom_size = (static_cast<uint32_t>(footerBytes[28]) << 24) | (static_cast<uint32_t>(footerBytes[29]) << 16) | (static_cast<uint32_t>(footerBytes[30]) << 8) | static_cast<uint32_t>(footerBytes[31]);
            footer.magic = NewMagicValue;
        } else {
            fh->seekg(size - LegacyFooterSize, std::ios::beg);
            std::vector<uint8_t> legacyFooterBytes(LegacyFooterSize);
            fh->read(reinterpret_cast<char*>(legacyFooterBytes.data()), LegacyFooterSize);
            
            std::string legacyMagic(reinterpret_cast<char*>(legacyFooterBytes.data() + 20), 8);
            if (legacyMagic != LegacyMagicValue) {
                throw CorruptFooterError();
            }
            
            footer.index_offset = (static_cast<uint64_t>(legacyFooterBytes[0]) << 56) | (static_cast<uint64_t>(legacyFooterBytes[1]) << 48) | (static_cast<uint64_t>(legacyFooterBytes[2]) << 40) | (static_cast<uint64_t>(legacyFooterBytes[3]) << 32) | (static_cast<uint64_t>(legacyFooterBytes[4]) << 24) | (static_cast<uint64_t>(legacyFooterBytes[5]) << 16) | (static_cast<uint64_t>(legacyFooterBytes[6]) << 8) | static_cast<uint64_t>(legacyFooterBytes[7]);
            footer.index_count = (static_cast<uint32_t>(legacyFooterBytes[8]) << 24) | (static_cast<uint32_t>(legacyFooterBytes[9]) << 16) | (static_cast<uint32_t>(legacyFooterBytes[10]) << 8) | static_cast<uint32_t>(legacyFooterBytes[11]);
            footer.record_count = (static_cast<uint32_t>(legacyFooterBytes[12]) << 24) | (static_cast<uint32_t>(legacyFooterBytes[13]) << 16) | (static_cast<uint32_t>(legacyFooterBytes[14]) << 8) | static_cast<uint32_t>(legacyFooterBytes[15]);
            footer.sparse_gap = (static_cast<uint32_t>(legacyFooterBytes[16]) << 24) | (static_cast<uint32_t>(legacyFooterBytes[17]) << 16) | (static_cast<uint32_t>(legacyFooterBytes[18]) << 8) | static_cast<uint32_t>(legacyFooterBytes[19]);
            footer.magic = LegacyMagicValue;
        }
    } else {
        throw CorruptFooterError();
    }

    std::vector<SparseIndexEntry> index = readIndex(*fh, footer);

    std::unique_ptr<bloom::Filter> bf;
    if (footer.magic == NewMagicValue && footer.bloom_offset > 0 && footer.bloom_size > 0) {
        bf = readBloom(*fh, footer);
    }

    return std::unique_ptr<File>(new File(path, std::move(fh), std::move(index), footer, std::move(bf), level, seq));
}

std::pair<record::Record, bool> File::Get(const std::vector<uint8_t>& key) {
    if (!fh_) {
        throw std::runtime_error("sstable closed");
    }

    if (bloom_ && !bloom_->MightContain(key)) {
        return {record::Record{}, false};
    }

    uint64_t startOffset = 0;
    uint64_t endOffset = footer_.index_offset;
    if (footer_.bloom_offset > 0) {
        endOffset = footer_.bloom_offset;
    }

    if (!index_.empty()) {
        auto it = std::lower_bound(index_.begin(), index_.end(), key, [](const SparseIndexEntry& entry, const std::vector<uint8_t>& k) {
            return entry.key < k;
        });

        if (it == index_.end() || it->key > key) {
            if (it != index_.begin()) {
                --it;
            }
        }

        startOffset = it->offset;
        auto next = it + 1;
        if (next != index_.end()) {
            endOffset = next->offset;
        }
    }

    uint64_t offset = startOffset;
    while (offset < endOffset) {
        try {
            auto [rec, nextOffset] = readRecordAt(*fh_, offset, endOffset);
            if (rec.key == key) {
                return {rec, !rec.is_deleted};
            }
            if (rec.key > key) {
                return {record::Record{}, false};
            }
            offset = nextOffset;
        } catch (const std::exception& e) {
            if (std::string(e.what()) == "EOF") {
                break;
            }
            throw;
        }
    }

    return {record::Record{}, false};
}

std::vector<record::Record> File::All() {
    if (!fh_) {
        throw std::runtime_error("sstable closed");
    }

    std::vector<record::Record> records;
    uint64_t offset = 0;
    uint64_t limit = footer_.index_offset;
    if (footer_.bloom_offset > 0) {
        limit = footer_.bloom_offset;
    }

    while (offset < limit) {
        try {
            auto [rec, nextOffset] = readRecordAt(*fh_, offset, limit);
            records.push_back(rec);
            offset = nextOffset;
        } catch (const std::exception& e) {
            if (std::string(e.what()) == "EOF") {
                break;
            }
            throw;
        }
    }
    return records;
}

} // namespace bigdb::sstable
